Darwin-Op modeled using Hinge2Joints

ad-daniel opened the pull request:
**Description**
Version of the Darwin-Op using Hinge2Joints instead of HingeJoints on ankle, pelvis and neck.

**Details**
The boundingObjects before and after overlap exactly (the robots are superposed here)
![darwin-op-vs-darwin-op-hinge2](https://user-images.githubusercontent.com/44834743/111292030-0daad980-8648-11eb-9795-6180497c4369.png)

properties before
![before](https://user-images.githubusercontent.com/44834743/111635229-9025cb80-87f7-11eb-92b8-aa397185af46.png)
properties after
![after](https://user-images.githubusercontent.com/44834743/111635246-9451e900-87f7-11eb-96bf-fe1300b867ce.png)

the change is due to the fact that having removed  intermediary solids, their mass was redistributed (and CoM computed from boundingObjects instead of enforced).

**Performance impact**
Scenario:
- 8 robots walking straight
- timestep 8 [ms]
- run on a (old) laptop (Geforce GT 730M). (Threads = 1, shadows disables, AA disabled)
- 2000 steps

World Metrics in `walk.wbt`::
|                            | solids | joints | geometries |
|----------------------------|--------|--------|------------|
| HingeJ (backlash = false)  | 265    | 160    | 305        |
| HingeJWB (backlash = 0)    | 265    | 160    | 305        |
| HingeJWB (backlash = 0.04) | 425    | 320    | 305        |
| Hinge2J (backlash = false)  | 225    | 120    | 305        |
| Hinge2JWB (backlash = 0)    | 225    | 120    | 305        |
| Hinge2JWB (backlash = 0.04) | 385    | 280    | 305        |

Performance in `walk.wbt`:

|                             | prePhysics(ms)  | physics(ms)     | postPhysics(ms) | mainRendering(ms) |
|-----------------------------|-----------------|-----------------|-----------------|-------------------|
| HingeJ (backlash = false)   | 655 [0.328]     | 31588 [15.794]  | 6000 [3.000]    | 10048 [5.024]     |
| HingeJWB (backlash = 0)     | 741 [0.371]     | 31766 [15.883]  | 7212 [3.606]    | 9615 [4.796]      |
| HingeJWB (backlash = 0.04)  | 1193 [0.597]    | 180843 [90.422] | 13639 [6.820]   | 8609 [4.294]      |
| Hinge2J (backlash = false)  | 511 [0.256]     | 17098 [8.549]   | 4702 [2.351]    | 4332 [4.194]      |
| Hinge2JWB (backlash = 0)    | 542 [0.271]     | 17140 [8.570]   | 5338 [2.669]    | 4339 [4.193]      |
| Hinge2JWB (backlash = 0.04) | 1017 [0.509]    | 127727 [63.864] | 11929 [5.965]   | 8474 [4.227]      |

It was obtained using the `--log-performance` argument. The values represent: TOTAL [AVG].

World Metrics in `kid.wbt`::
|                            | solids | joints | geometries |
|----------------------------|--------|--------|------------|
| HingeJ (backlash = false)  | 269    | 160    | 340        |
| HingeJWB (backlash = 0)    | 269    | 160    | 340        |
| HingeJWB (backlash = 0.04) | 429    | 320    | 340        |
| Hinge2J (backlash = false)  | 229    | 120    | 340        |
| Hinge2JWB (backlash = 0)    | 229    | 120    | 340        |
| Hinge2JWB (backlash = 0.04) | 389 | 280    | 340        |

Performance in `kid.wbt`:

|                             | prePhysics(ms)  | physics(ms)     | postPhysics(ms) | mainRendering(ms) |
|-----------------------------|-----------------|-----------------|-----------------|-------------------|
| HingeJ (backlash = false)   | 743 [0.372]     | 53651 [26.826]  | 6460 [3.230]    | 8762 [4.370]     |
| HingeJWB (backlash = 0)     | 806 [0.403]     | 52737 [26.369]  | 9595 [4.798]    | 7969 [3.975]      |
| HingeJWB (backlash = 0.04)  | 1270 [0.635]    | 255018 [127.509] | 17871 [8.936]   | 8121 [4.051]      |
| Hinge2J (backlash = false)  | 705 [0.353]     | 33502 [16.751]   | 5474 [2.737]    | 22664 [11.310]      |
| Hinge2JWB (backlash = 0)    | 762 [0.381]     | 33769 [16.885]   | 7742 [3.871]    | 20450 [10.200]      |
| Hinge2JWB (backlash = 0.04) | 1113 [0.557]    | 200702 [100.351] | 14967 [7.484]   | 7633 [3.807]      |


**Screenshots**
Walking (left is original, right uses hinge2)


https://user-images.githubusercontent.com/44834743/111640451-7f2b8900-87fc-11eb-8b08-6967335cf6b2.mp4



Standing 


https://user-images.githubusercontent.com/44834743/111640428-7a66d500-87fc-11eb-8c6b-320871d3ae1f.mp4

omichel commented on projects/robots/robotis/darwin-op/protos/Darwin-op-Hinge2.proto:
```diff
@@ -0,0 +1,1839 @@
+
+PROTO Darwin-op-Hinge2 [
+  field SFVec3f    translation      0 0 0                      # Is `Transform.translation`.
+  field SFRotation rotation         0 1 0 0                    # Is `Transform.rotation`.
```
See https://cyberbotics.com/doc/reference/proto-design-guidelines#translation-and-rotation
```suggestion
  field SFRotation rotation         0 0 1 0                    # Is `Transform.rotation`.
```

omichel commented on projects/robots/robotis/darwin-op/protos/Darwin-op-Hinge2.proto:
```diff
@@ -0,0 +1,1839 @@
+PROTO Darwin-op-Hinge2 [
+  field SFVec3f    translation      0 0 0                      # Is `Transform.translation`.
+  field SFRotation rotation         0 1 0 0                    # Is `Transform.rotation`.
+  field SFString   name             "DARwIn-OP-Hinge2"         # Is `Solid.name`.
```
```suggestion
  field SFString   name             "DARwIn-OP Hinge2"         # Is `Solid.name`.
```

omichel commented on projects/robots/robotis/darwin-op/protos/Darwin-op-Hinge2.proto:
```diff
@@ -0,0 +1,1839 @@
+  field MFNode     leftHandSlot     []                         # Extends the robot with new nodes in the left hand slot.
+  field MFNode     rightHandSlot    []                         # Extends the robot with new nodes in the right hand slot.
+#protected
+  hiddenField SFString _model       "DARwIn-OP-Hinge2"
```
```suggestion
  hiddenField SFString _model       "DARwIn-OP Hinge2"
```

omichel commented on projects/robots/robotis/darwin-op/protos/Darwin-op-Hinge2.proto:
```diff
@@ -0,0 +1,1839 @@
+# - camera
+# - realistic physics
+
+PROTO Darwin-op-Hinge2 [
```
I would rename it to "Darwin-opHinge2" to be consistent with other proto names.

omichel commented on projects/robots/robotis/darwin-op/protos/RobotisOp2Hinge2.proto:
```diff
@@ -0,0 +1,67 @@
+PROTO RobotisOp2Hinge2 [
+  field SFVec3f    translation     0 0 0                  # Is `Transform.translation`.
+  field SFRotation rotation        0 1 0 0                # Is `Transform.rotation`.
+  field SFString   name            "ROBOTIS OP2 HINGE2"   # Is `Solid.name`.
```
```suggestion
  field SFString   name            "ROBOTIS OP2 Hinge2"   # Is `Solid.name`.
```

omichel commented on projects/robots/robotis/darwin-op/worlds/darwin-op-hinge-vs-hinge2.wbt:
```diff
@@ -0,0 +1,38 @@
+#VRML_SIM R2021b utf8
```
I would rename this file `darwin-op_hinge_vs_hinge2.wbt` to be consistent with other worlds naming.

omichel commented on projects/robots/robotis/darwin-op/worlds/darwin-op-hinge-vs-hinge2.wbt:
```diff
@@ -0,0 +1,38 @@
+    "The ROBOTIS OP2 robot simulation model can be programmed using the ROBOTIS OP2 motions files and the ROBOTIS OP2 walk control library."
+  ]
+  title "ROBOTIS OP2 modeling comparison"
+  basicTimeStep 8
```
I would leave a time step of 16 ms if it works here.
```suggestion
  basicTimeStep 16
```

ad-daniel replied:
Since it doesn't have backlash enabled by default it can be 16 yes.

omichel reviewed (changes requested):
Thank you.
I believe we should a line in the ROBOTIS OP2 documentation (user guide) to mention this version with backlash.

ad-daniel commented:
> I believe we should a line in the ROBOTIS OP2 documentation (user guide) to mention this version with backlash.

The addition of the backlash field has been done in the [other](https://github.com/cyberbotics/webots/pull/2881) PR as it's more relevant there as this is just a modified version that uses Hinge2. Or do you reckon I should activate backlash on this sample world?

As for how to document this variant, I'm not sure what's the best way to go about it without creating too much confusion. 

- I've added a subsection `RobotisOp2Hinge2 variant` to the PROTO description where it just mentions what's different about this variant and where to find it
- I've added the sample world.

Should I mention more than this?

omichel reviewed (approved):
Very good. Thank you!
